# Advanced Dataframes

In [1]:
import pandas as pd
import numpy as np

np.random.seed(123)

In [2]:
# Create list of values for names column.

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# Randomly generate arrays of scores for each student for each subject.
# Note that all the values need to have the same length here.

math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))


In [3]:
# Construct the DataFrame using the above lists and arrays.

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades,
                   'classroom': np.random.choice(['A', 'B'], len(students))})


## Creating Dataframes

From lists and dictionaries

There are several ways to create dataframes, we've already seen how we can create a dataframe from a dictionary:

In [4]:
pd.DataFrame({'A': [1,2,3], 'B': [4,5,6]})

A  B
0  1  4
1  2  5
2  3  6

The keys in the passed dictionary will be the column names, and the values are the data points that make up each column.

We can also create dataframes from a 2d data structure, either a numpy array or a list of lists:

- a 2d data structure is made by using '[[]]'

In [5]:
pd.DataFrame([[1,2,3],[4,5,6]])

0  1  2
0  1  2  3
1  4  5  6

In [6]:
array = np.array([[1,2,3], [4,5,6]])
pd.DataFrame(array, columns=['a', 'b', 'c' ])

a  b  c
0  1  2  3
1  4  5  6

## From PyDataset

In some of the exercises, you'll need to load several datasets using the pydataset library. (If you get an error when trying to run the import below, use pip to install the pydataset package.) When the instructions say to load a dataset from PyDataset, you will need to do the following:

In [7]:
from pydataset import data

Running this code snippet will show you the valuable information doc on the dataset:

data(df_string_name, show_doc=True)


Running this code snippet will load the dataset for use as a pandas DataFrame:

In [9]:
df = data(df_string_name)

NameError: name 'df_string_name' is not defined

There are 757 available datasets using pydataset. Running the following code snippet in a cell will return a DataFrame with all of your options:

In [10]:
data()

dataset_id                                             title
0    AirPassengers       Monthly Airline Passenger Numbers 1949-1960
1          BJsales                 Sales Data with Leading Indicator
2              BOD                         Biochemical Oxygen Demand
3     Formaldehyde                     Determination of Formaldehyde
4     HairEyeColor         Hair and Eye Color of Statistics Students
..             ...                                               ...
752        VerbAgg                  Verbal Aggression item responses
753           cake                 Breakage Angle of Chocolate Cakes
754           cbpp                 Contagious bovine pleuropneumonia
755    grouseticks  Data on red grouse ticks from Elston et al. 2001
756     sleepstudy       Reaction times in a sleep deprivation study

[757 rows x 2 columns]

In [11]:
# Load the dataset and store it in the variable mpg.

from pydataset import data
mpg = data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [12]:
data('mpg', show_doc=True) 

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




In [13]:
from env import host, user, password
url = f'mysql+pymysql://{user}:{password}@{host}/employees'

ImportError: cannot import name 'host' from 'env' (/Users/deangelobowen/Desktop/numpy-pandas-visualization-exercises/env.py)

In [14]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', url)

NameError: name 'url' is not defined

It is common to have longer SQL queries that we want to read into python, and an example of how we might break a query into several lines is below:

In [15]:
sql = '''
SELECT
    emp_no,
    first_name,
    last_name
FROM employees
WHERE gender = 'F'
LIMIT 100
'''

employees = pd.read_sql(sql, url)
employees.head()

NameError: name 'url' is not defined

In [16]:
query = '''
SELECT
    t.title as title,
    d.dept_name as dept_name
FROM titles t
JOIN dept_emp USING (emp_no)
JOIN departments d USING (dept_no)
LIMIT 100
'''

title_dept = pd.read_sql(query, url)
title_dept.head()

NameError: name 'url' is not defined

# Indexing and Subsetting 

Like the pandas Series object, the pandas DataFrame object supports both position- and label-based indexing using the indexing operator [].

I will demonstrate concrete examples of indexing using the indexing operator [] alone and with the .loc and .iloc attributes below.

In [17]:
# Choose only two columns for my subset.

df[['name', 'classroom']]

name classroom
0     Sally         A
1      Jane         B
2     Suzie         A
3     Billy         B
4       Ada         A
5      John         B
6    Thomas         A
7     Marie         A
8    Albert         A
9   Richard         A
10    Isaac         B
11     Alan         A

In [18]:
# I can pass a boolean Series to the indexing operator as a selector.

bools = df.name.str.startswith('A')
bools

0     False
1     False
2     False
3     False
4      True
5     False
6     False
7     False
8      True
9     False
10    False
11     True
Name: name, dtype: bool

In [19]:
df[bools]

name  math  english  reading classroom
4      Ada    77       92       98         A
8   Albert    92       62       87         A
11    Alan    92       62       72         A

# .loc

We can use the .loc attribute to select specific rows AND columns by index label. The index label can be a number, but it can also be a string label. This method offers a lot of flexibility! The .loc attribute's indexing is inclusive and uses an index label, not position.

df.loc[row_indexer, column_indexer]

In [20]:
# Select all the rows and a subset of columns; notice the inclusive behavior of the indexing.

df.loc[:, 'math':'reading']

# Rows are first ':' (start at the beginning)
# then specify a range of columns. Between math and reading 'math:reading'

math  english  reading
0     62       85       80
1     88       79       67
2     94       74       95
3     98       96       88
4     77       92       98
5     79       76       93
6     82       64       81
7     93       63       90
8     92       62       87
9     69       80       94
10    92       99       93
11    92       62       72

In [21]:
# I can use a boolean Series as a selector with .loc, too, but I can choose rows and columns.

df.loc[bools, 'name': 'reading']

name  math  english  reading
4      Ada    77       92       98
8   Albert    92       62       87
11    Alan    92       62       72

# .iloc

We can use the .iloc attribute to select specific rows and colums by index position. .iloc does not accept a boolean Series as a selector like .loc does. It takes in integers representing index position and is NOT inclusive.

df.iloc[row_indexer, column_indexer]

In [22]:
# Notice the exclusive behavior of the indexing.

df.iloc[:3]

name  math  english  reading classroom
0  Sally    62       85       80         A
1   Jane    88       79       67         B
2  Suzie    94       74       95         A

In [23]:
df.iloc[:3, 1:3]

math  english
0    62       85
1    88       79
2    94       74

# Aggregating

.agg

The .agg method lets us specify a way to aggregate a series of numerical values. We pass an aggregate function or list of functions to the method that we want applied to a Series.

In [24]:
df.reading.agg('min')

67

While on the surface this seems pretty simple, .agg is capable of providing more detailed aggregations:

In [25]:
df[['english', 'reading', 'math']].agg(['mean', 'min', 'max'])

english  reading       math
mean  77.666667     86.5  84.833333
min   62.000000     67.0  62.000000
max   99.000000     98.0  98.000000

# .groupby

The .groupby method is used to create a grouped object, which we can then apply an aggregation on. For example, if we wanted to know the highest math grade from each classroom:

In [26]:
df.groupby('classroom').math.max()

classroom
A    94
B    98
Name: math, dtype: int64

In [27]:
df.groupby('classroom').math.agg(['min', 'mean', 'max'])

min    mean  max
classroom                  
A           62  82.625   94
B           79  89.250   98

We can group by multiple columns as well. To demonstrate, we'll create a boolean column named passing_math, then group by the combination of our new feature, passing_math, and the classroom and calculate the average reading grade and the number of individuals in each subgroup.

Let's break this problem down and code it step-by-step.

np.where

First, we can create the new passing_math column using a handy NumPy function called np.where. It will allow us to base the new column values on whether the values in an existing column, math, meet a condition.

np.where(condition, this_where_True, this_where_False)


In [28]:
# Create the new column based on an existing column.

df['passing_math'] = np.where(df.math < 70, 'failing', 'passing')
df

name  math  english  reading classroom passing_math
0     Sally    62       85       80         A      failing
1      Jane    88       79       67         B      passing
2     Suzie    94       74       95         A      passing
3     Billy    98       96       88         B      passing
4       Ada    77       92       98         A      passing
5      John    79       76       93         B      passing
6    Thomas    82       64       81         A      passing
7     Marie    93       63       90         A      passing
8    Albert    92       62       87         A      passing
9   Richard    69       80       94         A      failing
10    Isaac    92       99       93         B      passing
11     Alan    92       62       72         A      passing

Next, we will group by the passing_math and classroom columns and use the .agg method to calculate the average reading grade and the number of students.

In [29]:
grade_groups = df.groupby(['passing_math', 'classroom']).reading.agg(['mean', 'count'])
grade_groups

mean  count
passing_math classroom                  
failing      A          87.000000      2
passing      A          87.166667      6
             B          85.250000      4

In [30]:
# I can even clean up my columns to make my calculations clearer.

grade_groups.columns = ['avg_reading_grade', 'count_of_students']
grade_groups


avg_reading_grade  count_of_students
passing_math classroom                                      
failing      A                  87.000000                  2
passing      A                  87.166667                  6
             B                  85.250000                  4

# Takeaways:

We can interpret this output as there being 2 students failing math in classroom A with an average reading grade of 87, 6 students passing math in classroom A with an average reading grade of 87.16, and 4 students passing math in classroom B with an average reading grade of 85.25.

.transform

The .transform method can be used to produce a series with the same length of the original dataframe where each value represents the aggregation from the subgroup resulting from the .groupby.

For example, if we wanted to know the average math grade for each classroom and add this data back to our original dataframe:

In [31]:
df.assign(avg_math_score_by_classroom=
          df.groupby('classroom').math.transform('mean'))

name  math  english  reading classroom passing_math  \
0     Sally    62       85       80         A      failing   
1      Jane    88       79       67         B      passing   
2     Suzie    94       74       95         A      passing   
3     Billy    98       96       88         B      passing   
4       Ada    77       92       98         A      passing   
5      John    79       76       93         B      passing   
6    Thomas    82       64       81         A      passing   
7     Marie    93       63       90         A      passing   
8    Albert    92       62       87         A      passing   
9   Richard    69       80       94         A      failing   
10    Isaac    92       99       93         B      passing   
11     Alan    92       62       72         A      passing   

    avg_math_score_by_classroom  
0                        82.625  
1                        89.250  
2                        82.625  
3                        89.250  
4                        82.625  
5                        89.250  
6                        82.625  
7                        82.625  
8                        82.625  
9                        82.625  
10                       89.250  
11                       82.625

In [32]:
df.groupby('classroom').reading.describe().T


classroom         A          B
count       8.00000   4.000000
mean       87.12500  85.250000
std         8.88719  12.392874
min        72.00000  67.000000
25%        80.75000  82.750000
50%        88.50000  90.500000
75%        94.25000  93.000000
max        98.00000  93.000000

# Merging and Joining

Pandas provides several ways to combine dataframes together. We will look at two of them below:

pd.concat

This function takes in a list or dictionary of Series or DataFrame objects and joins them along a particular axis, row-wise axis=0 or column-wise axis=1.

## For example, concat with a list of two DataFrames
pd.concat([df1, df2], axis=0)

- When your list contains at least one DataFrame, a DataFrame is returned.

- When concatenating only Series objects row-wise, axis=0, a Series is returned.

- When concatenating Series or DataFrames column-wise, axis=1, a DataFrame is returned.

## Default is set to row-wise concatenation using an outer join.
pd.concat(objs, axis=0, join='outer')


In [33]:
df1 = pd.DataFrame({'a': [1, 2, 3]})
df2 = pd.DataFrame({'a': [4, 5, 6]})

df1

a
0  1
1  2
2  3

In [34]:
df2

a
0  4
1  5
2  6

In [35]:
pd.concat([df1, df2])

a
0  1
1  2
2  3
0  4
1  5
2  6

Note that the indices are preserved on the resulting dataframe; we could set the ignore_index parameter to True if we wanted these to be sequential.

In [36]:
concat_df1 = pd.concat([df1, df2], ignore_index=True)
concat_df1

a
0  1
1  2
2  3
3  4
4  5
5  6

In [37]:
concat_df2 = pd.DataFrame({'b': [1, 2, 3, 4, 5, 6]})
concat_df2

b
0  1
1  2
2  3
3  4
4  5
5  6

In [38]:
pd.concat([concat_df1, concat_df2], axis=1)

a  b
0  1  1
1  2  2
2  3  3
3  4  4
4  5  5
5  6  6

# .merge

This method is similar to a SQL join. Here's a cool read making a comparison between the two, if you're interested:

## df.merge default settings for commonly used parameters.

left_df.merge(right_df, how='inner', on=None, left_on=None, right_on=None, left_index=False, right_index=False, indicator=False)

- How does changing the default argument of the how parameter change my resulting DataFrame?

- how == Type of merge to be performed.

- how=left: use only keys from left frame, similar to a SQL left outer join; preserve key order.

- how=right: use only keys from right frame, similar to a SQL right outer join; preserve key order.

- how=outer: use union of keys from both frames, similar to a SQL full outer join; sort keys lexicographically.

- how=inner: use intersection of keys from both frames, similar to a SQL inner join; preserve the order of the left keys.


In [39]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [40]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

The .merge method will allow us to specify left_on and right_on to indicate the columns that are the keys used to merge the dataframes together.

In addition, the how keyword argument is used to define what type of JOIN we want to do; as we saw above, inner is the default setting.

For demonstration purposes, I'm also going to set the indicator parameter to True, which will create a column indicating whether the merge key appears in the left_only, right_only or both DataFrames.

In [41]:
# Perform an outer join specifying the left and right DataFrame keys.

users.merge(roles, left_on='role_id',
            right_on='id', how='outer', indicator=True)


id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

Notice that we have duplicate column names in the resulting dataframe. By default, pandas will add a suffix of _x to any columns in the left dataframe that are duplicated, and _y to any columns in the right dataframe that are duplicated. I can clean up my columns if I want to; one way would be to use method chaining, which it demonstrated below:

In [42]:
(users.merge(roles, 
            left_on='role_id', 
            right_on='id', 
            how='outer')
    .drop(columns='role_id')
    .rename(columns={'id_x': 'id', 
                     'name_x': 'employee',
                     'id_y': 'role_id',
                     'name_y': 'role'}
            )
)


id employee  role_id       role
0  1.0      bob      1.0      admin
1  2.0      joe      2.0     author
2  3.0    sally      3.0   reviewer
3  4.0     adam      3.0   reviewer
4  5.0     jane      NaN        NaN
5  6.0     mike      NaN        NaN
6  NaN      NaN      4.0  commenter

# Reshaping

We will talk about reshaping operations in more detail when we discuss tidy data, but for now we will focus on a couple of common operations that can be used to summarize our data by different subgroups.

pd.crosstab

For an example of .crosstab, we will count the number of students passing math in each classroom.

In [43]:
# We will use our student grades DataFrame, df.

df

name  math  english  reading classroom passing_math
0     Sally    62       85       80         A      failing
1      Jane    88       79       67         B      passing
2     Suzie    94       74       95         A      passing
3     Billy    98       96       88         B      passing
4       Ada    77       92       98         A      passing
5      John    79       76       93         B      passing
6    Thomas    82       64       81         A      passing
7     Marie    93       63       90         A      passing
8    Albert    92       62       87         A      passing
9   Richard    69       80       94         A      failing
10    Isaac    92       99       93         B      passing
11     Alan    92       62       72         A      passing

- pd.crosstab function counts the number of occurances in each subgroup
- this is like each unique combination of classroom AND whether or not the studen is passing math

In [44]:
pd.crosstab(df.passing_math, df.classroom)

classroom     A  B
passing_math      
failing       2  0
passing       6  4

In [45]:
pd.crosstab(df.passing_math, df.classroom, margins=True)

classroom     A  B  All
passing_math           
failing       2  0    2
passing       6  4   10
All           8  4   12

In [46]:
pd.crosstab(df.passing_math, df.classroom, normalize=True).round(3)


classroom         A      B
passing_math              
failing       0.167  0.000
passing       0.500  0.333

# .pivot_table

Here we use the .pivot_table method to create our summary. This method produces output similar to an excel pivot table. We must supply 3 things here:

- which values will make up the rows (the index)
- which values will make up the columns
- the values we are aggregating
- an aggregation method (aggfunc); if we can omit this, and mean will be used by default

## For an example using the pivot_table method, we'll calculate the average math grade for the combination of classroom and passing_math status: 

In [47]:
df.pivot_table(index='classroom', columns='passing_math', values='math')

passing_math  failing    passing
classroom                       
A                65.5  88.333333
B                 NaN  89.250000

pd.DataFrame(df.groupby(['day', 'smoker','time']).timecount())

Here we'll create a dataframe that represents various orders at a restaurant.

In [56]:
n = 40

orders = pd.DataFrame({
    'drink': np.random.choice(['Tea', 'Water', 'Water'], n),
    'meal': np.random.choice(['Curry', 'Yakisoba Noodle', 'Pad Thai'], n),
    'appetizer': np.random.choice(['Karaage','Edamame','Potato Salad'], n)
})

orders.sample(10)


drink             meal     appetizer
38  Water         Pad Thai  Potato Salad
13  Water         Pad Thai       Edamame
6   Water  Yakisoba Noodle       Karaage
25  Water  Yakisoba Noodle  Potato Salad
29  Water  Yakisoba Noodle       Karaage
7   Water         Pad Thai  Potato Salad
20    Tea  Yakisoba Noodle       Karaage
30  Water         Pad Thai  Potato Salad
5   Water         Pad Thai       Karaage
12    Tea            Curry       Edamame

# .map

## The .map method lets us use a dictionary to calculate the total price for an order; then I can save my calculations to a new column named bill. Let's do this step-by-step.

In [57]:
# Create a dictionary of prices for drinks and meals.

prices = {
    'Yakisoba Noodle': 9,
    'Curry': 11,
    'Pad Thai': 10,
    'Tea': 2,
    'Water': 0,
    'Karaage': 5,
    'Edamame': 3,
    'Potato Salad': 4
}


In [58]:
"""
Match the values in the 'drink' and 'meal' columns with the values in the 'prices' dictionary 
and perform the specified calculation. Save this calculation to a new column named 'bill'.
"""

orders['bill'] = orders.drink.map(prices) + orders.meal.map(prices) + orders.appetizer.map(prices)

orders.sample(25)


drink             meal     appetizer  bill
9     Tea         Pad Thai  Potato Salad    16
24  Water         Pad Thai       Edamame    13
32    Tea  Yakisoba Noodle       Edamame    14
1   Water            Curry       Karaage    16
8   Water            Curry       Karaage    16
31    Tea  Yakisoba Noodle       Edamame    14
5   Water         Pad Thai       Karaage    15
6   Water  Yakisoba Noodle       Karaage    14
21  Water  Yakisoba Noodle       Edamame    12
16    Tea         Pad Thai  Potato Salad    16
14  Water         Pad Thai       Edamame    13
11  Water  Yakisoba Noodle       Karaage    14
26  Water         Pad Thai       Karaage    15
33  Water  Yakisoba Noodle       Karaage    14
27  Water  Yakisoba Noodle  Potato Salad    13
23    Tea         Pad Thai       Edamame    15
10  Water            Curry       Edamame    14
39    Tea            Curry       Edamame    16
18  Water            Curry       Karaage    16
38  Water         Pad Thai  Potato Salad    14
37  Water         Pad Thai       Edamame    13
35  Water         Pad Thai       Karaage    15
3   Water            Curry  Potato Salad    15
0     Tea  Yakisoba Noodle       Edamame    14
36  Water            Curry  Potato Salad    15

In [64]:
pd.crosstab(orders.drink, orders.meal)

meal   Curry  Pad Thai  Yakisoba Noodle
drink                                  
Tea        2         3                5
Water      7        11               12

In [65]:
pd.crosstab(orders.drink, orders.meal, normalize=True, margins=True)


meal   Curry  Pad Thai  Yakisoba Noodle   All
drink                                        
Tea    0.050     0.075            0.125  0.25
Water  0.175     0.275            0.300  0.75
All    0.225     0.350            0.425  1.00

In [66]:
orders.pivot_table(index='drink', columns='meal', values='bill')


meal       Curry   Pad Thai  Yakisoba Noodle
drink                                       
Tea    16.000000  15.666667        14.800000
Water  15.142857  13.909091        13.083333

In [68]:
orders.groupby(['drink', 'meal']).bill.mean()


drink  meal           
Tea    Curry              16.000000
       Pad Thai           15.666667
       Yakisoba Noodle    14.800000
Water  Curry              15.142857
       Pad Thai           13.909091
       Yakisoba Noodle    13.083333
Name: bill, dtype: float64

# Transposing

In [69]:
df.describe().T

count       mean        std   min    25%   50%    75%   max
math      12.0  84.833333  11.134168  62.0  78.50  90.0  92.25  98.0
english   12.0  77.666667  13.371158  62.0  63.75  77.5  86.75  99.0
reading   12.0  86.500000   9.643651  67.0  80.75  89.0  93.25  98.0

In [70]:
df.T

0        1        2        3        4        5        6   \
name            Sally     Jane    Suzie    Billy      Ada     John   Thomas   
math               62       88       94       98       77       79       82   
english            85       79       74       96       92       76       64   
reading            80       67       95       88       98       93       81   
classroom           A        B        A        B        A        B        A   
passing_math  failing  passing  passing  passing  passing  passing  passing   

                   7        8        9        10       11  
name            Marie   Albert  Richard    Isaac     Alan  
math               93       92       69       92       92  
english            63       62       80       99       62  
reading            90       87       94       93       72  
classroom           A        A        A        B        A  
passing_math  passing  passing  failing  passing  passing